File: atcoder/abc/go/abc217/abc217_G/abc217_G.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)
var wrtr = bufio.NewWriterSize(os.Stdout, 10_000_000)
var rdr = bufio.NewScanner(os.Stdin)
func gs() string  { rdr.Scan(); return rdr.Text() }
func gi() int     { i,e := strconv.Atoi(gs()); if e != nil {panic(e)}; return i }
func gi2() (int,int) { return gi(),gi() }
func ia(m int) []int { return make([]int,m) }
func powmod(a,e,mod int) int { res, m := 1, a; for e > 0 { if e&1 != 0 { res = res * m % mod }; m = m * m % mod; e >>= 1 }; return res }
func makefact(n int,mod int) ([]int,[]int) {
	fact,factinv := make([]int,n+1),make([]int,n+1)
	fact[0] = 1; for i:=1;i<=n;i++ { fact[i] = fact[i-1] * i % mod }
	factinv[n] = powmod(fact[n],mod-2,mod); for i:=n-1;i>=0;i-- { factinv[i] = factinv[i+1] * (i+1) % mod }
	return fact,factinv
}
const MOD int = 998244353
func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
	defer wrtr.Flush()
	infn := ""; if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
	if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = bufio.NewScanner(f) }
	rdr.Split(bufio.ScanWords); rdr.Buffer(make([]byte,1024),1_000_000_000)
	// PROGRAM STARTS HERE
	N,M := gi2()
	ways := ia(N+1)
	biggestset := 1 + (N-1) / M
	smallestset := biggestset-1
	numbiggest := N - smallestset * M
	numsmallest := M - numbiggest
	fact,factinv := makefact(5000,MOD)
	// Ignore the non-empty requirement for ways
	for i:=1;i<=N;i++ {
		if biggestset > i { ways[i] = 0; continue }
		waysperbig := fact[i] * factinv[i-biggestset] % MOD
		waysbig := powmod(waysperbig,numbiggest,MOD)
		wayspersmall := fact[i] * factinv[i-smallestset] % MOD
		wayssmall := powmod(wayspersmall,numsmallest,MOD)
		ways[i] = waysbig * wayssmall % MOD
	}
	for i:=1;i<=N;i++ {
		lans,s := ways[i],-1
		for j:=i-1;j>0;j-- {
			adder := fact[i] * factinv[j] % MOD * factinv[i-j] % MOD * ways[j] % MOD * s
			lans = (lans + MOD + adder) % MOD
			s *= -1 
		}
		lans = lans * factinv[i] % MOD
		fmt.Fprintln(wrtr,lans)
	}
}

File: abc/python/abc217/abc217_G.py

import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

MOD = 998244353

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    N,M = gis()
    ways = [0] * (N+1)
    biggestset = 1 + (N-1) // M
    smallestset = biggestset-1
    numbiggest = N - smallestset * M
    numsmallest = M - numbiggest

    fact = [1] * 5001; factinv = [1] * 5001
    for i in range(1,5001) : fact[i] = fact[i-1] * i % MOD
    factinv[5000] = pow(fact[5000],MOD-2,MOD)
    for i in range(4999,-1,-1) : factinv[i] = factinv[i+1] * (i+1) % MOD

    ## Ignore the non-empty requirement for ways
    for i in range(1,N+1) :
        if biggestset > i : 
            ways[i] = 0
        else :
            waysperbig = fact[i] * factinv[i-biggestset] % MOD
            waysbig = pow(waysperbig,numbiggest,MOD)
            wayspersmall = fact[i] * factinv[i-smallestset] % MOD
            wayssmall = pow(wayspersmall,numsmallest,MOD)
            ways[i] = waysbig * wayssmall % MOD

    ans = [0] * (N+1)
    for i in range(1,N+1) :
        lans = ways[i]; s = -1 
        for j in range(i-1,0,-1) :
            adder = fact[i] * factinv[j] % MOD * factinv[i-j] % MOD * ways[j] % MOD * s
            lans = (lans+adder) % MOD
            s *= -1
        ans[i] = lans * factinv[i] % MOD
    ansstr = "\n".join([str(x) for x in ans[1:]])
    print(ansstr)

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

